### PR TITLE
Custom emissions reporting measure

### PR DESCRIPTION
**Pull request by bonnema**

### CODE REVIEW

**Comment by nllong** [comment]
sweet, I'll take a look today!

File: lib/measures/add_ems_emissions_reporting/measure.rb [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+# start the measure
+class AddEMSEmissionsReporting < OpenStudio::Measure::ModelMeasure
+  # human readable name
+  def name
+    return 'Add EMS to Report Emissions'
+  end
+
+  # human readable description
+  def description
+    return 'This measure reports emissions based on user-provided eGrid subregion.'
```
Comment (nllong) [inline lib/measures/add_ems_emissions_reporting/measure.rb:49]: Can we make this clear what type of carbon emissions? I think this is Annual historical or future co2 right?

File: lib/measures/add_ems_emissions_reporting/measure.rb [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+  def modeler_description
+    return 'This measure uses the EnergyPlus Energy Management System to log and report emissions based on user provided eGrid subregion.'
+  end
+
+  # define the arguments that the user will input
+  def arguments(model)
+    args = OpenStudio::Measure::OSArgumentVector.new
+
+    # eGrid subregion
+    sub_regn = OpenStudio::Measure::OSArgument.makeStringArgument('sub_regn', true)
```
Comment (nllong) [inline lib/measures/add_ems_emissions_reporting/measure.rb:62]: I don't see a reason to make the argument name short... make it `subregion`

File: lib/measures/add_ems_emissions_reporting/measure.rb [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+    sub_regn.setDisplayName('eGrid subregion')
+    sub_regn.setDescription('eGrid subregion. Options are: AZNM, CAMX, ERCT, FRCC, MROE, MROW, NEWE, NWPP, NYST, RFCE, RFCM, RFCW, RMPA, SPNO, SPSO, SRMV, SRMW, SRSO, SRTV, and SRVC')
+    sub_regn.setDefaultValue('RMPA')
+    args << sub_regn
+
+    # future year
+    fut_year = OpenStudio::Measure::OSArgument.makeIntegerArgument('fut_year', true)
```
Comment (nllong) [inline lib/measures/add_ems_emissions_reporting/measure.rb:69]: change to `future_year`

File: lib/measures/add_ems_emissions_reporting/measure.rb [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+    fut_year.setDisplayName('Future Year')
+    fut_year.setDescription('Future Year. Options are: 2020 to 2050 in two year increments')
+    fut_year.setDefaultValue(2030)
+    args << fut_year
+
+    # historical year
+    his_year = OpenStudio::Measure::OSArgument.makeIntegerArgument('his_year', true)
```
Comment (nllong) [inline lib/measures/add_ems_emissions_reporting/measure.rb:76]: change to `historical_year`

File: lib/measures/add_ems_emissions_reporting/measure.rb [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+
+  # define what happens when the measure is run
+  def run(model, runner, usr_args)
+    super(model, runner, usr_args)
+
+    # use the built-in error checking
+    return false unless runner.validateUserArguments(arguments(model), usr_args)
+
+    # assign the user inputs to variables
+    sub_regn = runner.getStringArgumentValue('sub_regn', usr_args)
```
Comment (nllong) [inline lib/measures/add_ems_emissions_reporting/measure.rb:93]: see above

File: lib/measures/add_ems_emissions_reporting/measure.rb [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+    sch_file.setInterpolatetoTimestep(false)
+    sch_file.setMinutesperItem(60)
+
+    # add EMS sensor for schedule file
+    sch_sens = OpenStudio::Model::EnergyManagementSystemSensor.new(model, 'Schedule Value')
+    sch_sens.setKeyName("#{sub_regn} #{fut_year} Hourly Emissions Sch")
+    sch_sens.setName('Hour_Sen')
+
+    # add whole-building electricity sensor
+    sch_sens = OpenStudio::Model::EnergyManagementSystemSensor.new(model, 'Facility Total Purchased Electricity Energy')
```
Comment (nllong) [inline lib/measures/add_ems_emissions_reporting/measure.rb:136]: nice 💥

File: lib/measures/add_ems_emissions_reporting/measure.rb [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+    sch_sens.setKeyName('Whole Building')
+    sch_sens.setName('Elec_Sen')
+
+    ems_prgm = OpenStudio::Model::EnergyManagementSystemProgram.new(model)
+    ems_prgm.setName('Emissions_Calc_Prgm')
+    fut_data.each {|r| ems_prgm.addLine("SET fann = #{r['co2_rate']}") if r['subregion'] == sub_regn && r['year'] == fut_year}
+    his_data.each {|r| ems_prgm.addLine("SET hann = #{r['co2_rate']}") if r['subregion'] == sub_regn && r['year'] == his_year}
+    ems_prgm.addLine('SET hval = Hour_Sen')
+    ems_prgm.addLine('SET elec = Elec_Sen')
+    ems_prgm.addLine('SET mult = 1000 * 60 * 60') # J to kWh (1000J/kJ * 60hr/min * 60 min/sec)
```
Comment (nllong) [inline lib/measures/add_ems_emissions_reporting/measure.rb:146]: noice

**Review by nllong — CHANGES_REQUESTED** [review]
looking sweet. Looks like this is just the annual average emissions (historical and future), so maybe we want to update the measure name to be more clear.

Few small tweaks suggested on the measure argument names.

File: lib/measures/add_ems_emissions_reporting/measure.rb (hunk 1 above)
Comment (bonnema) [reply lib/measures/add_ems_emissions_reporting/measure.rb:49]: > Can we make this clear what type of carbon emissions? I think this is Annual historical or future co2 right?

It is (1) historical annual CO2, (2) future annual CO2, and (3) future hourly CO2.

Comment (nllong) [reply lib/measures/add_ems_emissions_reporting/measure.rb:49]: gotcha, thanks!

**Review by nllong — APPROVED** [review]
looked great! I don't think we want to merge this down right now until we get approval from the rest of the BTO Data Tools team.

**Comment by bonnema** [comment]
I say we wait to merge until we get all the data files.

File: lib/measures/add_ems_emissions_reporting/measure.rb [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+    # future year
+    future_year = OpenStudio::Measure::OSArgument.makeIntegerArgument('future_year', true)
+    future_year.setDisplayName('Future Year')
+    future_year.setDescription('Future Year. Options are: 2020 to 2050 in two year increments')
+    future_year.setDefaultValue(2030)
+    args << future_year
+
+    # historical year
+    historical_year = OpenStudio::Measure::OSArgument.makeIntegerArgument('historical_year', true)
+    historical_year.setDisplayName('Historial Year')
```
Comment (nllong) [inline lib/measures/add_ems_emissions_reporting/measure.rb:77]:
```suggestion
    historical_year.setDisplayName('Historical Year')
```

File: lib/measures/add_ems_emissions_reporting/measure.rb [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+    historical_year.setDescription('Historial Year. Options are: 2007, 2009, 2010, 2012, 2014, 2016, 2018, and 2019.')
```
Comment (nllong) [inline lib/measures/add_ems_emissions_reporting/measure.rb:78]:
```suggestion
    historical_year.setDescription('Historical Year. Options are: 2007, 2009, 2010, 2012, 2014, 2016, 2018, and 2019.')
```

**Review by nllong — CHANGES_REQUESTED** [review]

**Comment by nllong** [comment]
@tijcolem  - can you clean up some space of CI  por favor?

```
Caused by: hudson.plugins.git.GitException: Command "git init /srv/jenkins/git/openstudio-common-measures-gem/PR-80/8" returned status code 128:
stdout: 
stderr: error: copy-fd: write returned: No space left on device
fatal: cannot copy '/usr/share/git-core/templates/hooks/applypatch-msg.sample' to '/srv/jenkins/git/openstudio-common-measures-gem/PR-80/8/.git/hooks/applypatch-msg.sample': No space left on device

	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:2681)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:2611)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:2607)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommand(CliGitAPIImpl.java:1974)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$5.execute(CliGitAPIImpl.java:1046)

```

**Comment by nllong** [comment]
@bonnema  - any reason not to merge this down once the tests pass?

**Comment by bonnema** [comment]
> @bonnema - any reason not to merge this down once the tests pass?

I think the only outstanding item is the additional data files, but I think those can be added at a later date.

**Comment by nllong** [comment]
Ok. I’ll merge once the tests pass! We also need to socialize this more. Maybe we do a quick preso to our group?

Da: Eric Bonnema ***@***.***>
Data: giovedì, 16 settembre 2021 10:35
A: NREL/openstudio-common-measures-gem ***@***.***>
Cc: Long, Nicholas ***@***.***>, Review requested ***@***.***>
Oggetto: Re: [NREL/openstudio-common-measures-gem] Custom emissions reporting measure (#80)

CAUTION: This email originated from outside of NREL. Do not click links or open attachments unless you recognize the sender and know the content is safe.

@bonnema<https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fbonnema&data=04%7C01%7Cnicholas.long%40nrel.gov%7Caa071a072fd7422e9e8908d9792ffd51%7Ca0f29d7e28cd4f5484427885aee7c080%7C0%7C0%7C637674069294067578%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=cUfmcBLMrL%2FeEQfnul%2FQN4tqCCpJVwUvsw4LtTe%2F%2Br0%3D&reserved=0> - any reason not to merge this down once the tests pass?

I think the only outstanding item is the additional data files, but I think those can be added at a later date.

**Comment by bonnema** [comment]
> Ok. I’ll merge once the tests pass! We also need to socialize this more. Maybe we do a quick preso to our group?

Sure. Maybe a monthly meeting presentation or something? I know there was a larger group of folks interested in this (that might not necessarily be in our group) so there might be a better forum. I am good with whatever.

**Comment by tijcolem** [comment]
@nllong Disk space filled up as I recently scheduled more jobs on these CI runners.  I re-run this job now and should run fine as plenty of disk space.

**Comment by nllong** [comment]
> @nllong Disk space filled up as I recently scheduled more jobs on these CI runners. I re-run this job now and should run fine as plenty of disk space.

sweet, thanks @tijcolem

**Comment by nllong** [comment]
Thanks @bonnema and @tijcolem ... merging now that tests are passing.

**Pull request merged**